script/build.js

```diff
@@ -0,0 +1,140 @@
+var ejs = require('ejs');
+
+/**
+ * Returns the method namespace of `method`, which is
+ * the method name minus the root and the basename.
+ * @param {String} method
+ * @returns {String}
+ */
+
+function namespace(method) {
+	return method.split('.').slice(1, -1).join('.');
+}
+
+/**
+ * Adds all of the "parts" of `ns` to `memo` and returns it.
+ * Some namespaces have intermediate parts that have no methods
+ * on their own. This is problematic when trying to create all
+ * of our namespaces in javascript as the intermediate objects
+ * will be undefined and assigning to them will throw. This
+ * method helps us ensure that all intermediate namespaces
+ * will be created.
+ * @param {String[]} memo
+ * @param {String} ns
+ * @returns {String[]}
+ */
+
+function parts(memo, ns) {
+	return memo.concat(ns.split('.').map(function (_, i, arr) {
+		return arr.slice(0, i + 1).join('.');
+	}));
+}
+
+/**
+ * Reduce function that only adds `str` to `memo` if it
+ * does not already exist in the array.
+ * @param {String[]} memo
+ * @param {String} str
+ * @returns {String[]}
+ */
+
+function uniq(memo, str) {
+	if (memo.indexOf(str) === -1) {
+		memo.push(str);
+	}
+	return memo;
+}
+
+/**
+ * Returns a filter function that returns whether or not
+ * `method` is under the namespace `ns`.
+ * @param {String}
+ * @returns {Function}
+ */
+
+function inNamespace(ns) {
+	return function (method) {
+		return namespace(method) === ns;
+	};
+}
+
+/**
+ * Returns whether or not `arg` is a required argument.
+ * @param {Object} arg
+ * @returns {Boolean}
+ */
+
+function required(arg) {
+	return !parseInt(arg.optional, 10);
+}
+
+/**
+ * Render the client file
+ */
+
+ejs.renderFile(__dirname + '/index.ejs', {
+
+	/**
+	 * All available method info responses
+	 * @type {Object}
+	 */
+
+	methods: require('require-dir')('../src'),
+
+	/**
+	 * All of the possible namespaces
+	 * @type {String[]}
+	 */
+
+	get namespaces() {
+		return Object.keys(this.methods)
+		.map(namespace)
+		.reduce(parts, [])
+		.reduce(uniq, [])
+		.sort();
+	},
+
+	/**
+	 * Returns all of the method names that exist
+	 * under the namespace `ns`.
+	 * @param {String} ns
+	 * @returns {String[]}
+	 */
+
+	getMethodsByNamespace: function (ns) {
+		return Object.keys(this.methods).filter(inNamespace(ns));
+	},
+
+	/**
+	 * Returns all of the non-optional arguments for `method`.
+	 * @param {String} method
+	 * @returns {String[]}
+	 */
+
+	getRequiredArguments: function (method) {
+		return this.methods[method]
+		.arguments
+		.argument
+		.filter(required)
+		.map(function (arg) {
+			return arg.name;
+		});
+	},
+
+	/**
+	 * Returns the method basename of `method`, which is
+	 * everything after the last delimiter.
+	 * @param {String} method
+	 * @returns {String}
+	 */
+
+	basename: function (method) {
+		return method.split('.').pop();
+	}
+
+}, function (err, str) {
+	if (err) {
+		throw err;
+	}
+	process.stdout.write(str);
+});
```

Comment:
For future archeology’s sake, our meatspace conversation about whether to check this in or not boiled down to “eh, it’s a built file so let’s not check it in, but if we find out later it would be good to check it in, let’s do it.”